examples/proof_of_location.masm

```diff
@@ -0,0 +1,714 @@
+# Radians (scaled): -40057176.678431965, -75418939.14609599
+const.RIO_DE_JANEIRO_LAT_SIGNED=1
+const.RIO_DE_JANEIRO_LAT=40057176
+const.RIO_DE_JANEIRO_LON_SIGNED=1
+const.RIO_DE_JANEIRO_LON=75418939
+
+# Melbourne
+# Coordinates (degrees): -37.810663237404874, 144.88590927245338
+# Radians (scaled): -65992056.58554934, 252873615.65500936
+const.MELBOURNE_LAT_SIGNED=1
+const.MELBOURNE_LAT=65992056
+const.MELBOURNE_LON_SIGNED=0
+const.MELBOURNE_LON=252873615
+
+const.SCALING_FACTOR_LO=1000000000
+const.SCALING_FACTOR_HI=0
+
+# 2 * earth radius in kilometers.
+const.DOUBLE_EARTH_RADIUS_LO=12742
+const.DOUBLE_EARTH_RADIUS_HI=0
+
+# u64 UTILS
+# =================================================================================================
+
+#! Swaps two u64s.
+#! Inputs: [x_hi, x_lo, y_hi, y_lo]
+#! Outputs: [y_hi, y_lo, x_hi, x_lo]
+proc.u64_swap
+  movdn.3 movdn.3
+end
+
+#! Computes the absolute difference between two u64s.
+#!
+#! Inputs: [x_hi, x_lo, y_hi, y_lo]
+#! Outputs: [diff_abs_hi, diff_abs_lo]
+proc.u64_abs_difference
+  dup.3 dup.3 dup.3 dup.3
+  # => [x_hi, x_lo, y_hi, y_lo, x_hi, x_lo, y_hi, y_lo]
+  exec.u64::lt
+  # => [is_y_lt_x, x_hi, x_lo, y_hi, y_lo]
+  if.true
+    # If y < x we cannot compute y - x without underflow, so we swap them.
+    exec.u64_swap
+  end
+
+  # Now we can safely subtract y from x without underflow.
+  exec.u64::overflowing_sub
+  # Assert no underflow happened.
+  not assert
+end
+
+# i64 MATH
+# =================================================================================================
+
+#! Negates an i64 by computing its bitwise NOT and adding 1.
+#! Inputs: [x_hi, x_lo]
+#! Outputs: [x_hi, x_lo]
+#! NOTE: Does not handle the case of inverting the largest negative number.
+proc.i64_neg
+  u32not swap u32not swap
+  # => [x_inverted_hi, x_inverted_lo]
+  push.1.0 exec.u64::overflowing_add not assert
+  # => [-x_hi, -x_lo]
+end
+
+#! Returns 1 if the given number is signed (its most significant bit is set)
+#! without consuming its inputs, 0 otherwise.
+#! Inputs: [x_hi, x_lo]
+#! Outputs: [is_x_signed, x_hi, x_lo]
+proc.i64_is_signed
+  dup push.0x80000000 u32and u32shr.31
+  # => [is_signed, x_hi, x_lo]
+end
+
+#! Computes x * y of two i64s while respecting the result's sign.
+#! It computes abs(x) * abs(y) as u64s and then corrects the sign at the end.
+#! Inputs: [x_hi, x_lo, y_hi, y_lo]
+#! Outputs: [xy_hi, xy_lo]
+#! Locals:
+#! 0: is_x_signed
+#! 1: is_y_signed
+#! 2: is_result_signed
+proc.i64_mul.3
+  # Initialize result sign to 0.
+  push.0 loc_store.2
+  # => [x_hi, x_lo, y_hi, y_lo]
+
+  # Store whether the inputs are signed in locals.
+  exec.i64_is_signed loc_store.0
+  # => [x_hi, x_lo, y_hi, y_lo]
+  exec.u64_swap
+  # => [y_hi, y_lo, x_hi, x_lo]
+  exec.i64_is_signed loc_store.1
+  # => [y_hi, y_lo, x_hi, x_lo]
+
+  loc_load.1
+  # => [is_y_signed, y_hi, y_lo, x_hi, x_lo]
+  if.true
+    exec.i64_neg
+    # Set the sign of the result.
+    loc_load.2 push.1 u32xor loc_store.2
```

Comment:
I don't understand this line. Can't you just store `1` at slot `2`?

Reply:
You're right, thanks!